packages/itwin/tree-widget/src/components/trees/common/UsePerformanceReporting.tsx

```diff
@@ -0,0 +1,49 @@
+/*---------------------------------------------------------------------------------------------
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+
+import { useCallback, useEffect, useRef } from "react";
+
+import type { IModelConnection } from "@itwin/core-frontend";
+import type { Ruleset } from "@itwin/presentation-common";
+
+/** @internal */
+export interface UsePerformanceReportingProps {
+  treeIdentifier: string;
+  iModel: IModelConnection;
+  ruleset: Ruleset;
+  onPerformanceMeasured?: (featureId: string, duration: number) => void;
+}
+
+/** @internal */
+export interface UsePerformanceReportingResult {
+  onNodeLoaded?: (props: { node: string; duration: number }) => void;
+}
+
+/**
+ * Enables performance reporting for a tree component.
+ * @internal
+ */
+export function usePerformanceReporting(props: UsePerformanceReportingProps): UsePerformanceReportingResult {
+  const { treeIdentifier, iModel, ruleset, onPerformanceMeasured } = props;
+  const firstLoadRef = useRef(true);
+
+  useEffect(() => {
+    firstLoadRef.current = true;
+  }, [iModel, ruleset]);
```

Comment:
If I understand correctly dependency on `ruleset` and this `firstLoadRef` is needed for the special case we have in `ModelsTree` where we are using different node loader/ruleset for unfiltered/filtered tree. As this is a special case and from consumer perspective it's a single tree I think `ModelsTree` should report node load time only from unfiltered node loader.

Reply:
Removed the `ruleset` dependency and disabled performance reporting in `ModelsTree` when filtered state is used.